client/site-profiler/components/get-report-form/index.tsx

```diff
@@ -0,0 +1,233 @@
+import { FormLabel, FormInputValidation, Gridicon, Button } from '@automattic/components';
+import { CheckboxControl } from '@wordpress/components';
+import emailValidator from 'email-validator';
+import { translate } from 'i18n-calypso';
+import { FormEvent, useState } from 'react';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+import wp from 'calypso/lib/wp';
+import './styles.scss';
+
+type Errors = {
+	name?: string;
+	email?: string;
+	termsAccepted?: string;
+};
+
+export function GetReportForm( {
+	url,
+	token,
+	onClose,
+}: {
+	url?: string;
+	token?: string;
+	onClose: () => void;
+} ) {
+	const [ name, setName ] = useState( '' );
+	const [ email, setEmail ] = useState( '' );
+	const [ isTermsChecked, setIsTermsChecked ] = useState( false );
+	const [ errors, setErrors ] = useState< Errors | null >( null );
+	const [ responseError, setResponseError ] = useState< string | null >( null );
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const [ responseSuccess, setResponseSuccess ] = useState( false );
+
+	const handleSubmit = async ( e: FormEvent< HTMLFormElement > ) => {
+		e.preventDefault();
+
+		const formData = new FormData( e.currentTarget );
+		const errors = validateForm( formData );
+		if ( errors ) {
+			setErrors( errors );
+			return;
+		}
+		setErrors( null );
+		setIsSubmitting( true );
+
+		let response = null;
+		try {
+			response = await wp.req.post(
+				{
+					path: '/site-profiler/lead',
+					apiNamespace: 'wpcom/v2',
+				},
+				{
+					name: formData.get( 'name' ),
+					email: formData.get( 'email' ),
+					url,
+					hash: token,
+				}
+			);
+		} catch ( error ) {
+			setResponseError( error instanceof Error ? error.message : String( error ) );
+			return;
+		} finally {
+			setIsSubmitting( false );
+		}
+		if ( response.success ) {
+			setResponseSuccess( true );
+		}
+	};
+
+	const validateName = ( name: string ) => {
+		if ( ! name ) {
+			return translate( 'Name is required' );
+		}
+		return null;
+	};
+
+	const validateEmail = ( email: string, fullValidation = false ) => {
+		if ( ! email ) {
+			return translate( 'Email is required' );
+		}
+		if ( fullValidation && ! emailValidator.validate( email ) ) {
+			return translate( 'Please enter a valid email address' );
+		}
+		return null;
+	};
+
+	const validateTerms = ( terms: boolean ) => {
+		if ( ! terms ) {
+			return translate( 'Terms must be accepted' );
+		}
+		return null;
+	};
+
+	function validateForm( formData: FormData ) {
+		let errors = {};
+		let hasErrors = false;
+
+		let error = validateName( formData.get( 'name' ) as string );
+		if ( error ) {
+			errors = { ...errors, name: error };
+			hasErrors = true;
+		}
```

Comment:
We kind of repeat this block of code for all 3 validations, do you think is there a way to generalize this code? 

Rough  and not tested example: 
```ts
const validationMethods = {
  name: validateName,
  email: validateEmail,
  termsAccepted: validateTerms,
};

const errors = Object.keys(validationMethods).reduce(
  (errors, attributeName) => {
    return {
      ...errors,
      [attributeName]: validationMethods[attributeName](
        FormData.length(attributeName)
      ),
    };
  },
  {}
);

const hasErrors = Object.values(errors).find((error) => !!error);
```

Reply:
Good point, I just commented my considerations [in the previous comment](https://github.com/Automattic/wp-calypso/pull/90336#discussion_r1596496174). As commented, I think that for three fields is not worthy yet to extract common functionality, but if we add more fields, it's definitely a good approach